### ISSUE TITLE
Shape index executable relative path

### ISSUE THREAD

**Issue by jrpruit1** [mapbox/preprocessorcerer#16]
@rclark Found that this: https://github.com/mapbox/preprocessorcerer/blob/master/preprocessors/shp-index.preprocessor.js#L4 is the root of all of the problems when preprocessorcerer becomes a dependency. Do you know a better way to reference the executable file? Maybe use `require.main.filename` (https://nodejs.org/api/modules.html#modules_accessing_the_main_module) ?

**Comment by springmeyer** [comment]
Maybe you could do something like:

```js
var mapnik = require('mapnik');
var shapeindex = path.resolve(mapnik.module_path, '../bin', 'mapnik-shapeindex.js');
```

I originally added this in the hope to make getting this path easy, but we've not been using it yet. See https://github.com/mapnik/node-mapnik/commit/7b6172d05c0b47181f8a1517195ef3568f0427ff#diff-b887ad28de3c65e35a884ada43488ef2

**Comment by jrpruit1** [comment]
When I console.log the place where it is looking now, I get
```
/home/jake/Projects/Mapbox/unpacker/node_modules/preprocessorcerer/bin/node_modules/.bin/mapnik-shapeindex.js
```
What I actually want is
```
/home/jake/Projects/Mapbox/unpacker/.bin/mapnik-shapeindex.js
```
The result of `mapnik.module_path` is
```
/home/jake/Projects/Mapbox/unpacker/node_modules/mapnik/lib/binding/node-v11-linux-x64
```
Would something like `process.cwd()` work in this situation? I guess that makes it dependent upon where you call the unpacker script...

**Comment by rclark** [comment]
I think @springmeyer meant that `mapnik.module_path` will land you in a known place within mapnik's code. Since you then know where you are, you can back out a couple of levels and find the bin via 

```
var shapeindex = path.resolve(mapnik.module_path, '../../bin', 'mapnik-shapeindex.js');
```

Let's do this for now... but we don't really have any hard guarantee that the directory structure within node-mapnik is stable. @springmeyer I'm curious why this function ended up exposed via shell executable instead of as addon to the node module? Technical reasons or time constraints?

**Comment by rclark** [comment]
Fixed in #17

**Comment by springmeyer** [comment]
> I think @springmeyer meant that mapnik.module_path will land you in a known place within mapnik's code.

correct @rclark 

> but we don't really have any hard guarantee that the directory structure within node-mapnik is stable

good point: the backing out part should stay the same but there is no guarantee.

>  @springmeyer I'm curious why this function ended up exposed via shell executable instead of as addon to the node module? Technical reasons or time constraints?

Mapnik is optimized and simple because the API is structured around thread-safe, read only access to data. Because generating an index is write, it has always felt like too much complexity to have mapnik core API expose this. So, after planning a write API in mapnik and never doing it I realized that just using the command line tool is the best approach. The other viable alternative would be to just build indexes in memory, per thread if one is not pre-generated (this is what we do for GeoJSON). But doing it ahead of time is ideal. So if you are fine with the slight messyness of the command line tool I'm happy with this going forward.
